src/http-client/node-http2-client.ts

```diff
@@ -53,145 +112,95 @@ export class NodeHTTP2Client implements HTTPClient {
-    this.#idleTime = _options.idleTime;
-    // WIP: should be set to something different for streaming
-    this.#pathName = "/query/1";
+    // create a new session if necessary
+    if (session_rc.session === null || session_rc.session.closed) {
+      const http2_session_idle_ms = this.#http2_session_idle_ms;
+      const url = this.#url;
 
-    try {
-      this.internal = http2.connect(url);
-      this.internal.setTimeout(this.#idleTime, () => {
-        this.close();
-      });
-    } catch (error) {
-      throw new NetworkError(`Could not connect to Fauna`, { cause: error });
-    }
-  }
+      const new_session: ClientHttp2Session = http2
+        .connect(url)
+        .once("error", () => this.close())
+        .once("goaway", () => this.close());
 
-  close() {
-    this.internal.close();
-  }
+      new_session.setTimeout(http2_session_idle_ms, () => {
+        this.#closeForAll();
+      });
 
-  async request({
-    data: requestData,
-    headers: requestHeaders,
-    method,
-  }: SessionRequestOptions): Promise<HTTPResponse> {
-    let req: any;
-
-    return new Promise<HTTPResponse>((resolvePromise, rejectPromise) => {
-      const onResponse = (http2ResponseHeaders: any) => {
-        const status = Number(
-          http2ResponseHeaders[http2.constants.HTTP2_HEADER_STATUS]
-        );
-        let responseData = "";
-
-        // append response data to the data string every time we receive new data
-        // chunks in the response
-        req.on("data", (chunk: any) => {
-          responseData += chunk;
-        });
-
-        // Once the response is finished, resolve the promise
-        req.on("end", () => {
-          resolvePromise({
-            status,
-            body: responseData,
-            headers: http2ResponseHeaders,
-          });
-        });
-      };
+      session_rc.session = new_session;
+    }
 
-      try {
-        const httpRequestHeaders: any = {
-          ...requestHeaders,
-          [http2.constants.HTTP2_HEADER_PATH]: this.#pathName,
-          [http2.constants.HTTP2_HEADER_METHOD]: method,
-        };
+    session_rc.refs.add(this);
 
-        req = this.internal
-          .request(httpRequestHeaders)
-          .setEncoding("utf8")
-          .on("error", (error: any) => rejectPromise(error))
-          .on("response", onResponse);
-        req.write(JSON.stringify(requestData), "utf8");
-        req.end();
-      } catch (error) {
-        rejectPromise(error);
-      }
-    });
+    return session_rc.session;
   }
 }
+
+/**
+ * Wrapper to provide reference counting for sessions.
+ *
+ * @internal
+ */
+type SessionRC = {
+  /** A session to be shared among http clients */
+  session: ClientHttp2Session | null;
+  /**
+   * A Setof client references. We will only close the session when there are no
+   * references left. This cannot be a WeakSet, because WeakSets are not
+   * enumerable (can't check if they are empty).
+   */
+  refs: Set<NodeHTTP2Client>;
```

Comment:
as discussed offline we can move back to user a simple count in a patch to save space.